Add contributing guide and code of conduct

kidder opened the pull request:
## Proposed changes

- Add code of conduct
- Add contributing guide, including code review guidelines
- Update copyright year

closes #1108 

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [x] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder committed:
Make doygen markdown headings consistent with rest of docs

The old version gave warnings with doxygen 1.8.15.
Modifying this file required whitelisting for lines over 80 characters.

kidder committed:
Update copyright year

kidder committed:
Add code of conduct

kidder committed:
Add contributing guide

nilsvu commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+documentation](https://spectre-code.org/):
+
+* [Installing SpECTRE](installation.html)
+* [Running Status Checks Locally](travis_guide.html)
```
Can you link directly to the section? Also perhaps change the link to "Running tests locally"

kidder replied:
I was unable to figure out how; I would be happy to if someone figures out how to do it

moxcodes replied:
I gave it a quick try. Apparently the following works:
- enclose the subsection you want to link to in an html anchor:
```
<a name="local_travis_checks">
## How to perform the checks locally
</a>
```
- then link to the guide using the html section
` * [Running Status Checks Locally](travis_guide.html#local_travis_checks)`

kidder replied:
Thanks, that works!

nilsvu commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+
+#### Table Of Contents
+
+[Code of Conduct](#code-of-conduct)
```
These anchors don't work. Not sure how to fix

moxcodes replied:
I think all that needs to be done is to add the tags, for instance `## Code of Conduct  {#code-of-conduct}` on the code of conduct section, and similar on the remaining sections.

kidder replied:
I had to use html anchors; I get warnings with doxygen labels

kidder replied:
I've figured out how to use the standard doxygen labels

nilsvu commented on docs/MainSite/Main.md:
```diff
@@ -12,6 +12,8 @@ Table of Contents
 <li class="level1"><a href="tutorials.html">User tutorials</a></li>
 <li class="level1"><a href="dev_guide.html">Dev guide</a></li>
+<li class="level1"><a href="code_of_conduct.html">Code of Conduct</a></li>
+<li class="level1"><a href="contributing_to_spectre.html">Contributing to SpECTRE</a></li>
```
Could you add these links also to the dropdown menu

nilsvu commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+  with the suggested changes, which will be addressed in a subsequent
+  pull request by the author.
+- The SpECTRE core developer will merge the pull request when
+  satisfied.
```
This last sentence sounds a bit strange, I suggest you remove the `when satisfied`.

nilsdeppe replied:
Maybe something more like `... pull request once all comments have been addressed, the code passes CI, and all pull requests the pull request depends on have been merged.`?

geoffrey4444 replied:
And also when the commits are squashed? Maybe say something (ideally pointing to documentation) about how to squash commits?

nilsvu commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+  request author to squash their commits.
+- Once the commits are squashed, the author should ping the reviewers.
+- Once all reviewers have approved the pull request, someone should
+  ping the `@sxs-collaboration/spectre-core-devs`.
```
You could add here that the ping is not necessary if one of the reviewers is a member of that group already (unless you want the ping anyway, I would be fine with that). However, I don't think we should require a _second_ member of the core-devs group to add a review.

nilsdeppe replied:
I agree with @nilsleiffischer :)

nilsvu commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+* Use the present tense ("Add feature" not "Added feature")
+* Use the imperative mood ("Move cursor to..." not "Moves cursor to...")
+* Limit the first line to 72 characters or less
+* If needed, a blank second line followed by a more complete description
```
> If needed, add ..., also limited to 72 characters per line

nilsdeppe replied:
What is wrong with them being 80 chars? None of my commits have ever followed that, they're always 60 char first line, 80 later.

wthrowe replied:
I generally use 50 first line and "whatever the default emacs auto-fill width is" (apparently 70) for later lines.  (Roughly the 50/72 rule.)  I don't object to allowing longer lines than that, but we should keep lines no longer than 80 characters (excepting URLs, etc.).

nilsvu replied:
I think GitHub expects commit messages to be limited to 72 chars, at least it does some wrapping. Not sure, I can try to find out.

nilsvu replied:
> The reason for wrapping your description lines at the 72nd mark is that git log adds a padding of 4 blank spaces when displaying the commit message. To center the message perfectly, and make it read well on an 80-column terminal, you’d want to leave space for 4 more blank spaces at right side. That’s where 72 comes from.

https://medium.com/@preslavrachev/what-s-with-the-50-72-rule-8a906f61f09c

nilsvu commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+* Reference issues and pull requests liberally after the first line
+
+[good-first-issue]:https://github.com/sxs-collaboration/spectre/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22+sort%3Acomments-desc
+[help-wanted]:https://github.com/sxs-collaboration/spectre/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22+sort%3Acomments-desc
```
Remove sorting, or sort by something more useful than number of comments

kidder replied:
what sort would be more useful?  Other codes use number of comments  as a gauge of interest in the issue.

nilsvu replied:
More recent issue could be better as first issues.. not sure

nilsdeppe replied:
I'm also not sure. Just because an issue is old doesn't mean we don't care about it. I think most recent issue isn't great, but most recently update would probably be closer to what we want since an issue that we are interested in would probably also have fairly recent comments.

nilsvu reviewed (changes requested):
I think you summarised the review procedure very well here, just a few minor comments.

nilsvu commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+deadlines once the new design is finalized.
+
+If you would like feedback on a pull request prior to it being ready
+for formal reveiw, please label it with `in progress` and ping
```
Typo: review

moxcodes commented on docs/CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,107 @@
+discuss it informally and confidentially with one of the project
+leaders.
+
+- Be friendly and patient. During teleconferences and meetings,
```
To introduce this set of bullets, add an introductory sentence before the first bullet. Perhaps "More specific guidelines include:" or  "More specifically, we expect members of the SpECTRE community to:", or similar.

moxcodes commented on docs/MainSite/ReportingGuide.md:
```diff
@@ -0,0 +1,91 @@
+
+You will receive an email from the Code of Conduct Committee or
+project leader(s) you contacted, acknowledging receipt within 24 hours
+(and we will aim to respond much quicker than that). They will review
```
There are some inconsistent choices of first vs third person voice, here it is "we will aim to respond" and then "They will review", and this voice changing appears in other places later in the document. We don't have a preference of "we" vs "they", but the document should be internally consistent.

moxcodes commented on docs/MainSite/ReportingGuide.md:
```diff
@@ -0,0 +1,91 @@
+
+Once the Code of Conduct Committee or contacted project leader(s) have
+a complete account of the events they will make a decision on how to
+respond, possibly in consultation with the full Executive Committee
```
perhaps "the full SXS Executive Committee", for consistency with the list of executive committee members below.

moxcodes commented on docs/MainSite/ReportingGuide.md:
```diff
@@ -0,0 +1,91 @@
+
+The current members of the Code of Conduct Committee (and the only
+people that see emails sent to conduct@spectre-code.org) are:
+- TBD: Saul Teukolsky
```
Can the TBD become determined by the time this is posted?

kidder replied:
Yes, that is the plan

moxcodes commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+eXtreme Spacetimes (SXS) research program into the multi-messenger
+astrophysics of neutron star mergers, core-collapse supernovae, and
+gamma-ray bursts.  As such, almost all of the contributors to SpECTRE
+are members of SXS institutions, and a large amount of planning the
```
This sentence doesn't read that smoothly. perhaps "a large amount of planning the design of SpECTRE is done" can be rephrased to "a large amount of SpECTRE design is done".

moxcodes commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+Bugs are tracked as [GitHub
+issues](https://guides.github.com/features/issues/). When you are
+creating a bug report, please **include as many details as
+possible**. Fill out the required template, as the information it asks
```
"Fill out the required template, as the information it asks for helps us resolve issues faster." can be rephrased to "Fill out the required template; it prompts for information that helps us resolve issues faster."

geoffrey4444 replied:
"We ask that you fill out the required template, because it prompts for information that helps us to resolve open issues faster."

moxcodes commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+
+In the future, we hope that SpECTRE can be applied to problems across
+discipline boundaries, and that it can be a true community code.  At
+the present time, however, SpECTRE lacks the capability to solve
```
Not that this is incorrect, but we should remember to change "lacks the capability to solve realistic problems" when it is no longer true (hopefully soon?)

geoffrey4444 replied:
Reword more optimistically: "SpECTRE can't yet solve realistic problems, and broad overview documentation is still in an early, incomplete stage."

moxcodes commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+* **Describe the exact steps which reproduce the problem** in as many
+  details as possible. For example, start by explaining how you
+  started SpECTRE, e.g. the exact command you used in the terminal, or
+  the contents of the batch script you used.
```
"batch script" is often used in windows. Perhaps "submission script"?

nilsdeppe replied:
`batch` is the term used on supercomputers (e.g. SLURM has the `sbatch` command to submit a batch job to which you supply a (batch) job script). We could say `batch job script`.

moxcodes replied:
I'm happy with the more disambiguated `batch job script`.

moxcodes commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+* **Use a clear and descriptive title** for the issue to identify the
+  problem.
+* **Describe the exact steps which reproduce the problem** in as many
+  details as possible. For example, start by explaining how you
```
"in as many details as possible" can be rephrased to "in as much detail as possible"

moxcodes commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+looking through these `good first issue` and `help wanted` issues:
+
+* [Good first issues][good-first-issue] - issues which should only
+  require a few lines of code, and a test or two.
```
Some of the current issues labelled "good first issues" are far more complicated than this describes.

kidder replied:
then they should be relabeled 😛

moxcodes commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+* [Installing SpECTRE](installation.html)
+* [Running Status Checks Locally](travis_guide.html)
+
+### Pull Requests
```
The first sentence of this section should be something along the lines of "Code contributions to SpECTRE follow a (pull request model)[https://help.github.com/articles/about-pull-requests/]".

moxcodes commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+  days only the assigned reviewers can request changes, unless someone
+  believes the code is wrong.  Non-reviewers are allowed to make
+  comments, which the pull request author is encouraged to address,
+  but alternatively can create an issue with the suggested changes,
```
"... is encouraged to address, but alternatively can create an issue with the suggested changes, which will be addressed in a subsequent pull request by the author." could be rephrased to "... is encouraged to address. Alternatively, the author can create an issue with the suggested changes, assigned to themselves, which would then be addressed in a subsequent pull request by the author."

moxcodes commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+- If necessary, pull requests can also be discussed at one of the
+  weekly SpECTRE meetings.
+- If changes are requested, the author should fix all of them in one
+  or more fixup commits, and then ping the reviewers that the pull
```
"fixup commits" should link to the code review guidelines to explain what those are and the syntax.

hrueter replied:
Add something about the way to ping others, e.g. "by @mentioning them in a comment".
https://github.blog/2011-03-23-mention-somebody-they-re-notified/

moxcodes commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+- Once a pull request is updated, the reviewers should either request
+  further changes (removing the `updated` label) or tell the pull
+  request author to squash their commits.
+- Once the commits are squashed, the author should ping the reviewers.
```
"squashed" could link to a git guide.

geoffrey4444 replied:
I second adding a link to docs for squashing! My students have struggled a lot with squashing, and it was nontrivial for me when I was a new spectre developer, too. 

But maybe that needs to be a separate PR, if such docs don't exist yet?

wthrowe replied:
We don't need to write our own docs for everything.  We can just point to the documentation from the git developers: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

geoffrey4444 replied:
I'm happy with adding the link @wthrowe suggests

moxcodes reviewed (commented):
Group code review with @fmahebert and @markscheel .

nilsdeppe commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+
+This project and everyone participating in it is governed by the \ref
+code_of_conduct "SpECTRE Code of Conduct". By participating, you are
+expected to uphold this code. Please report unacceptable behavior to
```
"unacceptable behavior" -> "possible violations of the code of conduct" seems more appropriate to me

geoffrey4444 replied:
I agree with @nilsdeppe

nilsdeppe commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+
+* Use the present tense ("Add feature" not "Added feature")
+* Use the imperative mood ("Move cursor to..." not "Moves cursor to...")
+* Limit the first line to 72 characters or less
```
How do we plan on enforcing these limits?

wthrowe replied:
The "mechanical" ones (line length, blank lines) can be checked in CheckCommits.sh.  The ones requiring knowledge of grammar are, of course, harder.

geoffrey4444 replied:
I would say it's up to reviewers of a PR to ask to reword the commit message if necessary

nilsdeppe commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+* Use the imperative mood ("Move cursor to..." not "Moves cursor to...")
+* Limit the first line to 72 characters or less
+* If needed, a blank second line followed by a more complete description
+* Reference issues and pull requests liberally after the first line
```
I'm on the fence with this for a couple reasons:

1. People will get emails when things they're subscribed to get tagged.
2. If we ever have to move away from GitHub then I'm not sure we can guarantee numbers will line up. Given that we're hoping SpECTRE will survive for two decades or more it seems not impossible for version control to look very different then.

geoffrey4444 replied:
I'm on the fence but maybe lean the other way. In two decades, how likely will we need to refer to issues and pull requests (as opposed to the final commit history) from a decade ago?

nilsdeppe commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+
+The process described here has several goals:
+
+- Maintain SpECTRE's quality
```
maybe `quality` -> `code and documentation quality`?

nilsdeppe commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+- If any requested change is unclear to the author, they should ping
+  the reviewer and ask for clarification.  If needed, the author and
+  reviewers should talk to one another (in person, via Google hangout,
+  ...) to resolve any issues.
```
`...` -> `or some other verbal method if possible`

hrueter replied:
We think this sentence should be rephrased to
"Authors and reviewers are encouraged to talk to one another ..."

nilsdeppe commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+If you would like feedback on a pull request prior to it being ready
+for formal reveiw, please label it with `in progress` and ping
+whomever you wish to get feedback from.  As long as the `in progress`
+label remains, no one is obligated to review the pull request.
```
I think we should be stronger that `not obligated`. I always viewed the `in progress` as "Please don't look at this unless someone has explicitly asked you. It will be opened up for general review later, at which point everyone will have the normal X hours to provide comments and specific reviewers will be chosen."

nilsdeppe commented on docs/MainSite/ReportingGuide.md:
```diff
@@ -0,0 +1,91 @@
+eXtreme Spacetimes (SXS) research program into the multi-messenger
+astrophysics of neutron star mergers, core-collapse supernovae, and
+gamma-ray bursts.  As such, it falls under the oversight of the SXS
+Executive Committee whose current members are:
```
I'm not sure I agree with this last list. Some people on the list have no  or very minimal involvement with the code and therefore I don't think are in a position to provide oversight.

geoffrey4444 replied:
I think this statement is trying to say that SXS oversees the code, and I think that's fair, since for now everyone involved is an SXS member. Since SXS is governed by the exec committee, that's the group that takes oversight overall. I don't think this is intended to imply technical oversight, which is from the core developers, but instead it is meant to imply overall management. This is a good point to discuss, though!

kidder replied:
I meant this as managerial oversight, not technical as @geoffrey4444 says.  I've changed "oversight" to "managerial oversight"

nilsdeppe replied:
I understood that, but there are people on the list that again have as far as I have been able to tell had no managerial involvement in SpECTRE. I'm just not entirely sure what this paragraph and list are trying to achieve, which is probably what is confusing me.

hrueter commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+- When creating the pull request, the author should add the
+  appropriate labels.  If the pull request is not a `new design` or
+  `in progress`, the author may either assign two reviewers (if GitHub
+  suggested any) or add the `review help wanted` label.
```
Originally the "review help wanted" was intended for pull requests that "should be straightforward to review", which is not how it is described here.

Maybe another label with the name "reviewers wanted" would be more fitting.

geoffrey4444 replied:
"reviewers wanted" is good for this purpose. The other could simply be "help wanted".

nilsvu replied:
- For the "reviewers wanted" we could just use the GitHub feature that shows all PRs with less then 2 reviewers: Just select "Reviews" > "Review required" (
https://github.com/sxs-collaboration/spectre/issues?q=is%3Apr+is%3Aopen+review%3Arequired). Then we don't need a label for this at all. We can of course keep the label to draw more attention to PRs that lack reviewers.
- The existing "straightforward to review" label was intended to offload review work to non-"spectre-core-devs" before this new policy was worked out. I think we can either remove the label now, or rename it to something like "good first review" (if we need such a label, probably not actually).

kidder replied:
I would rather have a single label for `reviewers wanted`.  I think it is up to a potential reviewer to decide if a PR is straightforward for themselves to review.

hrueter commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+- Fix problems that are important to users
+- Engage the community in working toward the best possible code
+- Enable a sustainable system for SpECTRE's maintainers to review
+  contributions
```
We should add something here, saying that we want to reach science goals in a timely manner.

geoffrey4444 replied:
I think reaching science goals should be the first priority of our PR policy and thus the first bullet, above "maintain SpECTRE's quality"

hrueter commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+  but alternatively can create an issue with the suggested changes,
+  which will be addressed in a subsequent pull request by the author.
+- Assigned reviewers should submit their review in as timely a manner
+  as possible.
```
We should give some concrete time frame, e.g. 2 days for a pull request with 100 lines.
Ideally we strive to handle small pull requests within a few days.

kidder replied:
I'd rather not be concrete as every PR is different.  I think we just need to be more proactive in pinging reviewers

hrueter commented on docs/MainSite/ReportingGuide.md:
```diff
@@ -0,0 +1,91 @@
+The current members of the Code of Conduct Committee (and the only
+people that see emails sent to conduct@spectre-code.org) are:
+- TBD: Saul Teukolsky
+- TBD: Harald Pfieffer
```
`Pfeiffer`

hrueter reviewed (changes requested):
group review by @haraldp271 @nilsleiffischer @hrueter

geoffrey4444 commented on docs/CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,107 @@
+from [Django Project Code of
+Conduct](https://www.djangoproject.com/conduct/) which used text from
+the [Speak Up! Code of
+Conduct](http://web.archive.org/web/20141109123859/http://speakup.io/coc.html).
```
I'm very happy that we are adding the code of conduct to the spectre GitHub! I suggest mentioning here as well how to report potential violations.

kidder replied:
It is mentioned on lines 35-37.

geoffrey4444 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+
+SpECTRE is being developed in support of our collaborative Simulating
+eXtreme Spacetimes (SXS) research program into the multi-messenger
+astrophysics of neutron star mergers, core-collapse supernovae, and
```
No love for black holes?? :(

markscheel replied:
Yes, please add black holes.

geoffrey4444 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+discipline boundaries, and that it can be a true community code.  At
+the present time, however, SpECTRE lacks the capability to solve
+realistic problems, and broad overview documentation is incomplete.
+Therefore, if you are not a member of SXS, but are interested in
```
Shouldn't this sentence be under "how to contribute"?

kidder replied:
no

geoffrey4444 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+Before creating bug reports, please **perform a
+[search](https://github.com/sxs-collaboration/spectre/issues)** to see
+if the problem has already been reported. If it has **and the issue is
+still open**, add a comment to the existing issue instead of opening a
```
add -> "please add"

geoffrey4444 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+new one.
+
+> **Note:** If you find a **Closed** issue that seems like it is the
+> same thing that you're experiencing, open a new issue and include a
```
open -> "please open"

geoffrey4444 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+
+Provide more context by answering these questions:
+
+* **Can you reproduce the problem in both debug and release mode?**
```
Please note or link here how to switch between modes

geoffrey4444 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+* **What is the name and version of the OS you're using**?
+* If possible (for SXS computers or HPC systems), a **path to a run
+  directory** that is accessible by SpECTRE core developers.
+
```
I suggest we suggest that users attempt to reproduce in docker as another convenient alternative, since that's a known environment to us

geoffrey4444 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+Please follow these steps to have your contribution considered by the
+maintainers:
+
+1. Follow all instructions in the template.
```
The checkboxes are ambiguous to me. Am I supposed to check them off myself when submitting, or do reviewers do this? Some can't be checked until Travis finishes, e.g."

markscheel replied:
I agree, the check boxes are ambiguous. Remove them?

kidder replied:
There are no checkboxes in this PR; please file an issue if you want to discuss the already existing template.

geoffrey4444 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+In general, longer pull requests take longer to review, with the time
+scaling exponentially with the number of lines changed.  Therefore if
+your pull request is too large, we may ask you to break it up into
+several smaller pull requests.
```
I would also suggest preemptively making pull requests as small as reasonable, as these are the easiest to review.

geoffrey4444 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+  there are at least two reviewers, the `review help wanted` label
+  should be removed.
+- If there are not two reviewers after two days, a SpECTRE core
+  developer will assign reviewers, removing the `review help wanted`
```
Link to the list of spectre core developers?

geoffrey4444 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+label remains, no one is obligated to review the pull request.
+
+
+### Git Commit Messages
```
Git Commit Message Guidelines

geoffrey4444 commented on docs/MainSite/Main.md:
```diff
@@ -74,8 +75,7 @@ eXtreme Spacetimes (SXS) research program into the multi-messenger
 astrophysics of neutron star mergers, core-collapse supernovae, and
 gamma-ray bursts.
```
No love for black holes??

kidder replied:
That is outside the scope of this PR

nilsvu replied:
If you change this sentence in the contributing guidelines, I suggest that for consistency you may as well change it here, too. For example:

> SpECTRE is being developed in support of our collaborative Simulating eXtreme Spacetimes (SXS) research program into the multi-messenger astrophysics of black-hole and neutron-star mergers, core-collapse supernovae, and gamma-ray bursts.

geoffrey4444 commented on docs/MainSite/ReportingGuide.md:
```diff
@@ -0,0 +1,91 @@
+people that see emails sent to conduct@spectre-code.org) are:
+- TBD: Saul Teukolsky
+- TBD: Harald Pfieffer
+- TBD: ???
```
I volunteer to be on this committee

kidder replied:
Thank you for volunteering, but I think the members of this committee should be people who are familiar with SpECTRE, but rarely code.

geoffrey4444 commented on docs/MainSite/ReportingGuide.md:
```diff
@@ -0,0 +1,91 @@
+SpECTRE is being developed in support of our collaborative Simulating
+eXtreme Spacetimes (SXS) research program into the multi-messenger
+astrophysics of neutron star mergers, core-collapse supernovae, and
+gamma-ray bursts.  As such, it falls under the oversight of the SXS
```
Black holes?

haraldp271 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+  developer will assign reviewers, removing the `review help wanted`
+  label.  If for some reason, no reviewers have been assigned after
+  three days, the author of the pull request should feel free to ping
+  `@sxs-collaboration/spectre-core-devs`
```
It would be useful to somewhere explain whether the people in spectre-core-devs can change.  Specifically, is it possible/planned that dutifully spectre-contributors and reviewers will eventually be promoted to core-devs? I hope that's the case, and stating whatever expectation or thoughts might be around this, would make the structure more transparent and welcoming, and might encourage contributors to put in more effort.

kidder replied:
Yes, spectre-core-devs can change

wthrowe commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+
+The following is a set of guidelines for contributing to SpECTRE,
+which is hosted in the [Simulating eXtreme Spacetimes
+Organization](https://github.com/sxs-collabortion) on GitHub.
```
Link doesn't work: collabortion -> collaboration.

kidder committed:
Add code of conduct and contributing guide to menu

wthrowe commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+This project and everyone participating in it is governed by the \ref
+code_of_conduct "SpECTRE Code of Conduct". By participating, you are
+expected to uphold this code. Please report unacceptable behavior to
+[conduct@spectre-code.org](mailto:conduct@spectre-code.org).
```
All the other mailto: links in this PR just allow Doxygen to autoformat the link.

wthrowe commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,286 @@
+following manner:
+- When creating the pull request, the author should add the
+  appropriate labels.  If the pull request is not a `new design` or
+  `in progress`, the author may either assign two reviewers (if GitHub
```
`new design` and `in progress` have not been mentioned before and their meaning isn't clear yet.  Perhaps just add "(discussed later)" or something to make it clear that understanding this clause is not necessary for following the rest of the text.

wthrowe commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,313 @@
+  the `reviewers wanted` label.  If for some reason, no reviewers have
+  been assigned after three days, the author of the pull request
+  should feel free to ping `@sxs-collaboration/spectre-core-devs`
+- Anyone can request changes within the first two days.  After two
```
I believe we previously agreed on max(2 days, [assigned reviews submitted] + 1 day)

wthrowe commented on tools/FileTestDefs.sh:
```diff
@@ -153,6 +153,7 @@ long_lines() {
               'Doxyfile.in$' \
               'containers/Dockerfile.travis$' \
               'docs/MainSite/Main.md' \
+              'docs/CONTRIBUTING.md' \
```
I don't think you need this.  Lines containing links are already whitelisted, but unfortunately will show up in the report if something else is wrong.  Running the checks on this file doesn't actually seem to report anything, though.

wthrowe commented on docs/CONTRIBUTING.md:
```diff
@@ -224,13 +224,14 @@ following manner:
-  changes, assigned to themselves, which would then be addressed in a
-  subsequent pull request by the author.
+- Anyone can request changes within the first two days (or within a
+  day after two approvals on the first day).  After two days only the
```
`max(2 days, [assigned reviews submitted] + 1 day)`, not `min(2 days, [assigned reviews submitted] + 1 day)`

nilsvu commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,315 @@
+
+\tableofcontents
+
+# Contributing to SpECTRE {#contributing_to_spectre_2}
```
Instead of having these 2 identical headlines, you can have the headline once, then the one sentence, then the `\tableofcontents`

kidder replied:
that does not work; I get multiple warnings about finding commands out of context

nilsvu commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,315 @@
+  create an issue with the suggested changes, assigned to themselves,
+  which would then be addressed in a subsequent pull request by the
+  author.
+- Assigned reviewers should submit their review in as timely a manner
```
Move up by one bullet point

nilsvu commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,315 @@
+  developer](#core-developers) will assign reviewers, removing
+  the `reviewers wanted` label.  If for some reason, no reviewers have
+  been assigned after three days, the author of the pull request
+  should feel free to ping `@sxs-collaboration/spectre-core-devs`
```
".. by @mentioning them in a comment to the pull request." (You explain the "ping" in a bullet point below, but better move that here, since it's the first occurrence.)

nilsvu commented on docs/DevGuide/Travis.md:
```diff
@@ -5,6 +5,10 @@ See LICENSE.txt for details.
 
+\tableofcontents
+
+# Testing SpECTRE with Travis CI {#travis_ci}
```
Also here you can merge the two headlines, perhaps have the headline followed by `\tableofcontents` and then the paragraph in this case

nilsvu commented on docs/MainSite/ReportingGuide.md:
```diff
@@ -0,0 +1,90 @@
+SpECTRE is being developed in support of our collaborative Simulating
+eXtreme Spacetimes (SXS) research program into the multi-messenger
+astrophysics of neutron star mergers, core-collapse supernovae, and
+gamma-ray bursts.  As such, it falls under the managerial oversight of
```
same sentence here

nilsvu commented on docs/config/header.html:
```diff
@@ -81,6 +81,16 @@
           </li>
+          <li style="min-width: 135px;">
+            <a href="code_of_conduct.html">
+              <i class="octicon octicon-gist-secret"></i><span>Code of Conduct</span>
```
`octicon-comment-discussion`

nilsvu commented on docs/config/header.html:
```diff
@@ -81,6 +81,16 @@
+          </li>
+          <li style="min-width: 135px;">
+            <a href="contributing_to_spectre.html">
+              <i class="octicon octicon-gist-secret"></i><span>Contributing Guide</span>
```
`octicon-git-pull-request`

nilsvu reviewed (commented):
Very minor, but if you the chance..

kidder commented:
@haraldp271 @markscheel @nilsdeppe @nilsleiffischer @wthrowe @moxcodes @geoffrey4444 @hrueter  thank you for your comments; I have squashed my fixups.

wthrowe reviewed (dismissed):
I'm not going to block on it, but is it possible to make the new contributing guide table of contents look less awful?  It currently lists, as an example:

How
Do
I
Submit
A
(Good)
Bug
Report?

Edit: spelling

kidder commented:
I figured out how to make the TOC look better, and will push an additional commit soon...

kidder committed:
Make doxygen html table of contents wider

Also untabify and other cleanup to css file.
Also whitelist css file from including license.

kidder commented:
@wthrowe please see if you like the contributing guide table of contents better

wthrowe reviewed (approved):
Yeah, that's much nicer.

nilsdeppe commented:
Restarted builds that timed out (GCC6,7 debug+release, clang release)

nilsdeppe commented:
Travis GCC6 release timed out again. Merging anyway.
